### ISSUE TITLE
Add support for EVCharger

### ISSUE THREAD

**Issue by freakent** [freakent/dbus-mqtt-devices#110]
### Version

0.6.6 (current)

### What version of VenusOS are you running on your GX device

3.21 (current release)

### Does this issue occur during installation or at runtime ?

Other

### Is there a Discussion thread for this issue? Please paste a link.

https://github.com/freakent/dbus-mqtt-devices/issues/32#issuecomment-1979792013

@dersch81

### What is the issue you are actually experiencing?

Users are using the driver to add EV Chargers to their system using a customised service description. With each new release of the driver they have to manually add the service description to their services.yml file. 

The service description currently in use needs refining before inclusion in the released version. 

```
# https://github.com/victronenergy/venus/wiki/dbus#evcharger
evcharger:
  ProductId:
    default: 65535 # 0xFFFF
  CustomName:
    default: "MyCharger"
    persist: true
  Ac/Energy/Forward:
    description: "wh"
    format: "{:.2f}W"
  Ac/L1/Power:
    description: "W"
    format: "{} W"
  Ac/L2/Power:
    description: "w"
    format: "{} W"
  Ac/L3/Power:
    description: "w"
    format: "{} W"
  Ac/Power:
    description: "w"
    format: "{} W"
  AutoStart:
    description: "w"
    default: 0
#    persist: true
  ChargingTime:
    description: "in seconds"
    format: "{} s"
  Current:
    description: "A"
    default: 0
#    persist: true
  EnableDisplay:
    description: ""
  MaxCurrent:
    description: "A"
    default: 0
#    persist: true   #wird nicht gespeichert reset nach disconnect/reconnect
  Mode:
    description: ""
    default: 0
  Model:
    description: "w"
    default: "AC22NS"
  Position:
    persist: true
    default: 0
    min: 0
    max: 1
    description: "0=OUTPUT; 1=INPUT"
#  Connected:
#    description: "w"
  Role:
    description: "w"
  SetCurrent:
    description: "w"
#  StartStop:
#    description: "w"
#    persist: true
  Status:
    description: "w"
    default: 1
```

**Comment by freakent** [comment]
Things to do
1) confirm the use of “persist: true” across all attributes
2) use the correct descriptions from the Victron documentation 
3) confirm why some fields are commented out entirely 
4) create a script to test/verify the configuration is working

**Comment by freakent** [comment]
With correct descriptions:
```
# https://github.com/victronenergy/venus/wiki/dbus#evcharger
evcharger:
  ProductId:
    default: 65535 # 0xFFFF
  CustomName:
    default: "MyCharger"
    persist: true
  Ac/Energy/Forward:
    description: "wh"
    format: "{:.2f}W"
  Ac/L1/Power:
    description: "L1 Power used (W)"
    format: "{} W"
  Ac/L2/Power:
    description: "L3 Power used (W)"
    format: "{} W"
  Ac/L3/Power:
    description: "L3 Power used (W)"
    format: "{} W"
  Ac/Power:
    description: "AC Power (W)"
    format: "{} W"
  AutoStart:
    description: "w"
    default: 0
#    persist: true
  ChargingTime:
    description: "in seconds"
    format: "{} s"
  Current:
    description: "A"
    default: 0
#    persist: true
  EnableDisplay:
    description: ""
  MaxCurrent:
    description: "A"
    default: 0
#    persist: true   #wird nicht gespeichert reset nach disconnect/reconnect
  Mode:
    description: ""
    default: 0
  Model:
    description: "w"
    default: "AC22NS"
  Position:
    persist: true
    default: 0
    min: 0
    max: 1
    description: "0=OUTPUT; 1=INPUT"
#  Connected:
#    description: "w"
  Role:
    description: "w"
  SetCurrent:
    description: "w"
#  StartStop:
#    description: "w"
#    persist: true
  Status:
    description: "w"
    default: 1
```

**Comment by freakent** [comment]
Settings - aka where to use persist=true

CustomName
AutoStart ??
MaxCurrent ??
Mode
Model
Position
Role

**Comment by dersch81** [comment]
Sure, i have a quite big NodeRed Flow for this to read via MQTT the data from the Wallbox (i'm using two) in that way and registered two EVcharger. I can also control the charging mode via VRM, because i have mapped the mode ID's between VenusOS and OpenWB. But should work with any other "open" Wallbox, too.  

![image](https://github.com/freakent/dbus-mqtt-devices/assets/13754891/84fcd431-c3bd-40da-a182-538ec67c12cf)

[EVCharger Flow 1.json](https://github.com/freakent/dbus-mqtt-devices/files/14524345/EVCharger.Flow.1.json)
[EVCharger Flow 2.json](https://github.com/freakent/dbus-mqtt-devices/files/14524346/EVCharger.Flow.2.json)

In VRM they appear like this:
![image](https://github.com/freakent/dbus-mqtt-devices/assets/13754891/d18c3180-d1e4-457f-ae80-60bca11db2b4)

Example of local Venus GUI:
![image](https://github.com/freakent/dbus-mqtt-devices/assets/13754891/c6a430aa-423f-4bc0-b7b2-c3c74cbcce53)

MQTT Explorer:
![Screenshot 2024-03-07 140542](https://github.com/freakent/dbus-mqtt-devices/assets/13754891/aee27af4-e15a-49c4-a352-687ed9f18494)

![Screenshot 2024-03-07 140701](https://github.com/freakent/dbus-mqtt-devices/assets/13754891/5abf9c84-97eb-414b-8ad7-a0c0d4e3e13f)


dbus-spy:
![image](https://github.com/freakent/dbus-mqtt-devices/assets/13754891/c8eab1f7-b9b6-4c0f-b544-d312961340c7)

![image](https://github.com/freakent/dbus-mqtt-devices/assets/13754891/16dcf255-239b-4b27-a232-df4ac92936bd)

**Comment by freakent** [comment]
In your nodered flows, what does Gate do?

**Comment by dersch81** [comment]
In that flow the gate is just preventing to flood the register with zero values. It opens only in case of >0. 

In general i'm using gates a lot to control main flows with subflows.

**Comment by dersch81** [comment]
> In that flow the gate is just preventing to flood the register with zero values. It opens only in case of >0.
> 
> In general i'm using gates a lot to control main flows with subflows.

I'm sorry my information about the gate was wrong. It is controlling the counting of charging time. When the current is >0 then it opens the gate and the timestamp payload is able to flow. If current is =0 then the gate is closed and time counting stopps. The next process is starting from time 0 again.

**Comment by freakent** [comment]
These are the attributes that need persist = true

Mode
SetCurrent
MaxCurrent
Position
AutoStart
EnableDisplay
CustomName 

![image](https://github.com/freakent/dbus-mqtt-devices/assets/124850/e31615f5-a44a-4517-93fd-24b4c9fe602e)

**Comment by dersch81** [comment]
persist = true just means to have the same state after restart, right?

**Comment by freakent** [comment]
Yes, which also affects the device specific setup screen on the GX device (as per my screenshots above). If the correct fields are not set as persist = true the setup screen will not work correctly. It might have been better to call this parameter "setting" instead of "persist" to more accurately affect what it does.

**Comment by freakent** [comment]
Fixed in v0.8.0.
